bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/welcome/NAWelcomeHomeHandler.java

```diff
@@ -97,96 +100,164 @@ protected NAWelcomeHome updateReadings() {
+            }
+        }
+        super.triggerChannelIfRequired(channelId);
+    }
+
+    private static Set<String> findDetectedObjectTypes(Optional<NAWelcomeEvent> eventOptional) {
+        Set<String> detectedObjectTypes = new TreeSet<>();
+        if(!eventOptional.isPresent()) {
+            return detectedObjectTypes;
+        }
+
+        NAWelcomeEvent event = eventOptional.get();
+
+        if(event.getPersonId() != null) {
+            detectedObjectTypes.add(CAMERA_EVENT_OPTION_HUMAN_DETECTED);
+        }
+
+        if(NAWebhookCameraEvent.EventTypeEnum.MOVEMENT.toString().equals(event.getType())) {
+            detectedObjectTypes.add(CAMERA_EVENT_OPTION_MOVEMENT_DETECTED);
+        }
+
+        List<NAWelcomeSubEvent> subEvents = event.getEventList();
+        for(NAWelcomeSubEvent subEvent: subEvents) {
+            String detectedObjectType = translateDetectedObjectType(subEvent.getType());
+            detectedObjectTypes.add(detectedObjectType);
+        }
+        return detectedObjectTypes;
+    }
+
+    private static String translateDetectedObjectType(NAWelcomeSubEvent.TypeEnum type) {
+        switch (type) {
+            case HUMAN: return CAMERA_EVENT_OPTION_HUMAN_DETECTED;
+            case ANIMAL: return CAMERA_EVENT_OPTION_ANIMAL_DETECTED;
+            case VEHICLE: return CAMERA_EVENT_OPTION_VEHICLE_DETECTED;
+            default: throw new IllegalArgumentException("Unknown detected object type!");
+        }
+    }
+
+    private Optional<String> findEventMessage() {
+        if(lastEvent.isPresent()) {
+            @Nullable String message = lastEvent.get().getMessage();
+            if (message != null) {
+                return Optional.of(message);
+            }
+
+            return findFirstSubEvent(lastEvent).map(NAWelcomeSubEvent::getMessage);
+        }
+        return Optional.empty();
+    }
+
     /**
      * Returns the Url of the picture
      *
      * @return Url of the picture or null
      */
-    protected String getSnapshotURL(NAWelcomeSnapshot snapshot) {
-        String result = null;
-
-        if (snapshot != null && snapshot.getId() != null && snapshot.getKey() != null) {
-            result = WELCOME_PICTURE_URL + "?" + WELCOME_PICTURE_IMAGEID + "=" + snapshot.getId() + "&"
-                    + WELCOME_PICTURE_KEY + "=" + snapshot.getKey();
-        } else {
-            logger.debug("Unable to build snapshot url for Home : {}", getId());
+    protected Optional<String> findSnapshotURL() {
+        if(lastEvent.isPresent()) {
+            @Nullable NAWelcomeSnapshot snapshot = lastEvent.get().getSnapshot();
+            if (snapshot == null) {
+                snapshot = findFirstSubEvent(lastEvent).map(NAWelcomeSubEvent::getSnapshot).orElse(null);
+            }
+
+            if (snapshot != null && snapshot.getId() != null && snapshot.getKey() != null) {
+                return Optional.of(WELCOME_PICTURE_URL + "?" + WELCOME_PICTURE_IMAGEID + "=" + snapshot.getId() + "&"
```

Comment:
Can you change this to a StringBuilder or a String.format("....") ?

Reply:
IntelliJ shows the warning "'StringBuilder' can be replaced with 'String'". I think to use StringBuilder isn't required anymore since Java 6, because the compiler optimizes it itself. It is more readable when the Strings are concat with "+" in this case (there are no conditions, just simple String concat).

Reply:
It would look like this, but IntelliJ would show the warning, so I won't commit that:

`                return Optional.of(new StringBuilder()
                        .append(WELCOME_PICTURE_URL)
                        .append('?')
                        .append(WELCOME_PICTURE_IMAGEID)
                        .append('=')
                        .append(snapshot.getId())
                        .append('&')
                        .append(WELCOME_PICTURE_KEY)
                        .append('=')
                        .append(snapshot.getKey())
                        .toString());`